x/market/keeper/swap.go

```diff
@@ -103,22 +108,27 @@ func (k Keeper) ComputeSwap(ctx sdk.Context, offerCoin sdk.Coin, askDenom string
 	terraPool := basePool.Add(terraPoolDelta)
 	lunaPool := cp.Quo(terraPool)
 
-	var offerPool sdk.Dec // base denom(usdr) unit
-	var askPool sdk.Dec   // base denom(usdr) unit
+	var offerPool sdk.Dec       // base denom(usdr) unit
+	var askPool ConstrainedPool // base denom(usdr) unit
 	if offerCoin.Denom != core.MicroLunaDenom {
 		// Terra->Luna swap
 		offerPool = terraPool
-		askPool = lunaPool
+		askPool = ConstrainedPool{Pool: lunaPool, MaxSize: sdk.NewDec(2000000000000)}
 	} else {
 		// Luna->Terra swap
 		offerPool = lunaPool
-		askPool = terraPool
+		askPool = ConstrainedPool{Pool: terraPool}
+
+	}
+	// Simply hack to ensure that swaps into askPool can not happen if it will exceed MaxSize
+	if (askPool.MaxSize != sdk.Dec{} && askPool.Pool.Add(retDecCoin.Amount).GT(askPool.MaxSize)) {
```

Comment:
If the intention is to prevent LUNC from having more than 2T supply, Isn't this supposed to be Supply + Amount > Max?

Reply:
With the current implementation, what happens if the pool resets or intentionally swaps around max to avoid max?

Reply:
As I understand it the askPool.Pool is just the size  of the  pool we are trading into at the time of the trade, so the logic should be ok with regards to just ensuring that the expected amount of "retdeccoin" does not exceed the max size on the constrained pool? So not really sure the constant product is an issue here? :)

Reply:
What I am saying is that even if the logic works, it is unlikely to accomplish the purpose.

Reply:
I am not opposed to having multiple levels of defense against token stability, but I think the contents related to tokenomics should go through more specialized research and be written in the most appropriate place so that there are no side effects.

Reply:
I disagree. The hard cap I have proposed has been both tested and would most likely have prevented the de-peg from happening in the first place. :) Also I think after giving it some thought that this is a setting we might not want to make configurable via governance params due to the risk of this configuration option being abused by the whales. So I think it should be hardcoded, whats your thoughts on that matter? :)

Reply:
My point is that just because "the code seems to work" doesn't mean that it will actually accomplish its purpose, as no comprehensive tests have been written and no scenario reviews have been written to see if this code will completely limit the supply.

With some adjustments to the Oracle/Market parameters, it seems that you can easily mint more than 2T supply. Let me try a test.

I think it's part of the system, even if it has unhealthy governance, and you have to follow the rules set by the system. Blockchain is also a human product after all, and imperfect humans face such problems in real life. This is not a problem that can be solved with code planted by one or two people.

Reply:
Also, I believe that adding such code in a network where instability has not been resolved due to the de-pegging event could cause additional confusion. I believe that if this code is to be added to the mainnet, it should be decided through governance after a more comprehensive discussion in the community. (Whether the community is healthy or not)

Reply:
> My point is that just because "the code seems to work" doesn't mean that it will actually accomplish its purpose, as no comprehensive tests have been written and no scenario reviews have been written to see if this code will completely limit the supply.
> 
> With some adjustments to the Oracle/Market parameters, it seems that you can easily mint more than 2T supply. Let me try a test.
> 
> I think it's part of the system, even if it has unhealthy governance, and you have to follow the rules set by the system. Blockchain is also a human product after all, and imperfect humans face such problems in real life. This is not a problem that can be solved with code planted by one or two people.

Actually I initially broke all our unit tests around the market module by targeting a 10B supply. Thats why I set it to 2T to avoid having  to fix all the other tests. So in a sense the test have been done, but it makes sense to implement a test to ensure that the feature is working as intended. :)

With regards to the params  this is why I am so interested in  some sort of "proposal bundle" feature that has some IFTT logic to stipulate successful outcomes for the "vote bundle" before its either passed or rejected. Because the more params we have, the easier the system  will become to manipulate if people are not diligent about  the consequences of changes they vote for.

Your right. Conways Law will ensure that. I do however see this more as a starting point and then we will use  the  governance layer to debate the final product and respect the vote, even if it means the feature will never get adopted. :)

Reply:
> Also, I believe that adding such code in a network where instability has not been resolved due to the de-pegging event could cause additional confusion. I believe that if this code is to be added to the mainnet, it should be decided through governance after a more comprehensive discussion in the community. (Whether the community is healthy or not)

In my view this has already been voted thru governance in proposal 3568 which states that there should only ever be 10B supply of LUNC in existence. Unfortunately it  failed to provide the IKEA guide to tell us how to easily get there :P

Reply:
I have implemented a unit test to ensure that the overflow constrain works as intended :)

Reply:
I can't find anything in [proposal 3568](https://ping.pub/terra-luna/gov/3568) that the total supply cannot exceed 10B.

Reply:
I think the agora link in the prop leads to the v2 DNS. Here is the right link: https://classic-agora.terra.money/t/burn-mechanism-on-transactions-1-2/43759 - "A Tax Burn mechanism is to be implemented on LUNC to reduce the Total Supply. Implement a Tax + Burn mechanism on each buy-sell transaction: 1.2% burn tax **This mechanism should be true until the total supply = 10 billion LUNC, after that, this mechanism is disabled and the total supply can never be changed**."

Reply:
This would indicate that we should just hard code the 10B cap into the swap logic, at least that how I interpret it :)